sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala

```diff
@@ -16,254 +16,459 @@
-          try {
-            if (inputBatch.numCols() > 0) {
-              Some(sortBatch(inputBatch))
-            } else {
-              Some(new ColumnarBatch(Array.empty, inputBatch.numRows()))
-            }
-          } finally {
-            inputBatch.close()
-          }
-        } else {
-          None
         }
       }
+    } else {
+      val splitIndexes = if (sortedOffset >= 0) {
+        sortedOffset until rows by targetRowCount
+      } else {
+        targetRowCount until rows by targetRowCount
+      }
+      // Get back the first row so we can sort the batches
+      val gatherIndexes = if (sortedOffset >= 0) {
+        // The first batch is sorted so don't gather a row for it
+        splitIndexes
+      } else {
+        Seq(0) ++ splitIndexes
+      }
 
-      private def sortBatch(inputBatch: ColumnarBatch): ColumnarBatch = {
-        val nvtxRange = initNvtxRange
-        try {
-          var outputTypes: Seq[DataType] = Nil
-          var inputTbl: Table = null
-          var inputCvs: Seq[GpuColumnVector] = Nil
-          try {
-            if (sortOrder.nonEmpty) {
-              inputCvs = SortUtils.getGpuColVectorsAndBindReferences(inputBatch, sortOrder)
-              inputTbl = new cudf.Table(inputCvs.map(_.getBase): _*)
-              outputTypes = sortOrder.map(_.child.dataType) ++
-                  GpuColumnVector.extractTypes(inputBatch)
-            } else if (inputBatch.numCols() > 0) {
-              inputTbl = GpuColumnVector.from(inputBatch)
-              outputTypes = GpuColumnVector.extractTypes(inputBatch)
-            }
-            val orderByArgs = getOrderArgs(inputTbl)
-            val startTimestamp = System.nanoTime()
-            val batch = doGpuSort(inputTbl, orderByArgs, outputTypes)
-            updateMetricValues(inputTbl, startTimestamp, batch)
-            batch
-          } finally {
-            inputCvs.foreach(_.close())
-            if (inputTbl != null) {
-              inputTbl.close()
+      val boundaries =
+        withResource(new NvtxRange("boundaries", NvtxColor.ORANGE)) { _ =>
+          withResource(ColumnVector.fromInts(gatherIndexes: _*)) { gatherMap =>
+            withResource(sortedTbl.gather(gatherMap)) { boundariesTab =>
+              // Hopefully this is minor but just in case...
+              memUsed += gatherMap.getDeviceMemorySize +
+                  GpuColumnVector.getTotalDeviceMemoryUsed(boundariesTab)
+              convertBoundaries(boundariesTab)
             }
           }
-        } finally {
-          nvtxRange.close()
         }
-      }
 
-      override def hasNext: Boolean = {
-        if (resultBatch.isDefined) {
-          true
+      withResource(sortedTbl.contiguousSplit(splitIndexes: _*)) { split =>
+        memUsed += split.map(_.getBuffer.getLength).sum
+        val stillPending = if (sortedOffset >= 0) {
+          withResource(new NvtxRange("Store Sorted Table", NvtxColor.DARK_GREEN)) { _ =>
```

Comment:
Is this block expected to take a significant amount of time?  I'm surprised to see an NVTX range here, as I'm seeing only some object creation and adding to a linked list.

Reply:
I think this range is mainly for visibility.

Reply:
I added it when debugging what was taking a lot of time, and never removed it.  It turns out that the time was being taken by the next range where we add things to pending, because pending will do a lazy compile of the comparator which I was trying to track down. I am happy to remove this one if we want to.

Reply:
Unless there's a compelling reason for them to remain, I'd like to see the extremely small NVTX ranges like this one removed since they can distort profiles and probably aren't that useful in practice.  If these tiny ranges serve a useful purpose, may be worth commenting so someone doesn't come along later and remove them.

Reply:
Like with metrics, and log lines, should we have levels for nvtx ranges? This is a separate discussion but since we are talking about it, I can spawn off an issue if that's interesting.

Reply:
If I am looking at NVTX everything is already debug level. The main reason to remove them is because they are so small all the time that showing them has no useful value. I just didn't remove them earlier when I was debugging, and Jason did the right thing to ask "why is this section of code taking long enough that we need to pay attention to it?" Because if it did take that long then we have some optimizing to do.